Add guidance for trusted publishing (#1261)
pypa/packaging.python.org: source/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows.rst

``push`` event.
This guide shows you how to publish a Python distribution
whenever a tagged commit is pushed.
It will use the `pypa/gh-action-pypi-publish GitHub Action`_ for

> publishing and `upload-artifact`_ and `download-artifact`_ actions 

Reviewer: That's a long sentence. Let's try splitting it.
```suggestion
publishing. It also uses `upload-artifact`_ and `download-artifact`_ actions 
```

   details of building platform specific projects. If you have binary
   components, check out :ref:`cibuildwheel`'s GitHub Action examples.

Configuring trusted publishing
==============================

This guide relies on PyPI's `trusted publishing`_ implementation to connect
to `GitHub Actions CI/CD`_. This is recommended for security reasons, since 
the generated tokens are created for each of your projects
individually and expire automatically. Otherwise you'll need to generate an

> `API token`_ or provide a username/password combination for both PyPI and

Reviewer: I wouldn't suggest using username and password for PyPI or TestPyPI. Instead, I'd only mention such a possibility in the context of publishing to third party indexes like devpi.

TestPyPI. 

Since this guide will demonstrate uploading to both
PyPI and TestPyPI, we'll need two trusted publishers configured. 

> The following steps will lead you through creating the "pending" publishers.

Reviewer: Let's mention that it's also possible to configure trust for pre-existing projects, if one is an owner.

Let's begin! 🚀

> 1. Go to https://pypi.org/manage/account/publishing/

Reviewer: We probably need a trailing period for style consistency here:
```suggestion
1. Go to https://pypi.org/manage/account/publishing/.
```

2. Fill in the name you wish to publish your new project under,

>    your repository data and the name of the release workflow file 

Reviewer: Since it's a tutorial, we shouldn't send people scroll the page back and forth. Instead, we have specific values in the examples and should explicitly tell the reader what to enter.

   components, check out :ref:`cibuildwheel`'s GitHub Action examples.

Configuring trusted publishing
==============================

This guide relies on PyPI's `trusted publishing`_ implementation to connect
to `GitHub Actions CI/CD`_. This is recommended for security reasons, since 
the generated tokens are created for each of your projects
individually and expire automatically. Otherwise you'll need to generate an
`API token`_ for both PyPI and TestPyPI. In case of publishing to third-party
indexes like :doc:`devpi <devpi:index>`, you will need to provide a
username/password combination.

Since this guide will demonstrate uploading to both
PyPI and TestPyPI, we'll need two trusted publishers configured. 
The following steps will lead you through creating the "pending" publishers
for your new project. However it is also possible to add `trusted publishing`_
to any pre-existing project, if you are its owner.

Let's begin! 🚀

1. Go to https://pypi.org/manage/account/publishing/.
2. Fill in the name you wish to publish your new project under,
   your GitHub username and repository name and 
   the name of the release workflow file under 

>    the ``.github/`` folder, see :ref:`workflow-definition`. 

Author: I only have [Anaconda](http://damnwidget.github.io/anaconda/) in use, that doesn't care about trailing spaces in a RST file... Can you suggest me anything?

>    the ``.github/`` folder, see :ref:`workflow-definition`. 

Reviewer: I think we should integrate pre-commit and use https://github.com/pre-commit/pre-commit-hooks#trailing-whitespace... I'm not sure about other projects. I know that Vim/NeoVim can be configured to show trailing spaces visually, there's probably some plugins for that. Most projects just use pre-commit for this sort of linters/formatters.

>    the ``.github/`` folder, see :ref:`workflow-definition`. 

Author: Thanks for suggesting, I've opened #1268 for that.

   details of building platform specific projects. If you have binary
   components, check out :ref:`cibuildwheel`'s GitHub Action examples.

Configuring trusted publishing
==============================

This guide relies on PyPI's `trusted publishing`_ implementation to connect
to `GitHub Actions CI/CD`_. This is recommended for security reasons, since 
the generated tokens are created for each of your projects
individually and expire automatically. Otherwise you'll need to generate an
`API token`_ for both PyPI and TestPyPI. In case of publishing to third-party
indexes like :doc:`devpi <devpi:index>`, you will need to provide a
username/password combination.

Since this guide will demonstrate uploading to both
PyPI and TestPyPI, we'll need two trusted publishers configured. 
The following steps will lead you through creating the "pending" publishers

> for your new project. However it is also possible to add `trusted publishing`_

Reviewer 2: Maybe we should use project with link to glossary: https://packaging.python.org/en/latest/glossary/#term-Project

> for your new project. However it is also possible to add `trusted publishing`_

Author: Can we use a relative link for that anchor?

> for your new project. However it is also possible to add `trusted publishing`_

Reviewer: Usually, glossary terms are linkable through the `:term:` role.
And I see it's available indeed: https://webknjaz.github.io/intersphinx-untangled/packaging.python.org/.

> for your new project. However it is also possible to add `trusted publishing`_

Reviewer: I'd still like that extra context so how about both a reference and a hint?
```suggestion
2. Fill in the name you wish to publish your new :term:`PyPI project <Project>` under (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),
```

to any pre-existing project, if you are its owner.

Let's begin! 🚀

1. Go to https://pypi.org/manage/account/publishing/.

> 2. Fill in the name you wish to publish your new project under,

Reviewer 2: Name of project? Name of distribution package?

* https://packaging.python.org/en/latest/glossary/#term-Project
* https://packaging.python.org/en/latest/glossary/#term-Distribution-Package

If using the terms as defined in glossary (with link) then it should be unambiguous.

> 2. Fill in the name you wish to publish your new project under,

Reviewer: I'd still like that extra context so how about both a reference and a hint?
```suggestion
2. Fill in the name you wish to publish your new :term:`PyPI project <Project>` under (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),
```

`API token`_ for both PyPI and TestPyPI. In case of publishing to third-party
indexes like :doc:`devpi <devpi:index>`, you may need to provide a
username/password combination.

Since this guide will demonstrate uploading to both
PyPI and TestPyPI, we'll need two trusted publishers configured. 
The following steps will lead you through creating the "pending" publishers
for your new :term:`PyPI project <Project>`.
However it is also possible to add `trusted publishing`_ to any
pre-existing project, if you are its owner.

   .. attention::

      If you followed earlier versions of this guide, you
      have created the secrets ``PYPI_API_TOKEN`` and ``TEST_PYPI_API_TOKEN``
      for direct PyPI and TestPyPI access. These are obsolete now and
      you should remove them from your GitHub repository and revoke
      them in your PyPI and TestPyPI account settings in case you are replacing your old setup with the new one.

Let's begin! 🚀

1. Go to https://pypi.org/manage/account/publishing/.
2. Fill in the name you wish to publish your new
   :term:`PyPI project <Project>` under
   (the ``name`` value in your ``setup.cfg`` or ``pyproject.toml``),

>    your GitHub username and repository name and

Reviewer 3: <!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**suggestion:**  Since users may add publishing from an org repository, telling them to add their own username may be misleading.
‏
```suggestion
   the GitHub repository owner's name and repository name and
```

Reviewer: How about adding `(org or user)` clarification?

>    the GitHub repository owner's name (org or user)
>    and repository name and the name of the release workflow file under

Reviewer: I think we should stick some commas here:
```suggestion
   the GitHub repository owner's name (org or user),
   and repository name, and the name of the release workflow file under
```

   the ``.github/`` folder, see :ref:`workflow-definition`.

>    Finally add the name of the GitHub Actions environment

Reviewer: This is a bit confusing but the term is GitHub Environment. There's no "GitHub Actions environments" but the GitHub Environments can be specified in the GitHub Actions job definitions which is something different.
```suggestion
   Finally, add the name of the GitHub Environment
```

   (``pypi``) we're going set up under your repository.
   Register the trusted publisher.
3. Now, go to https://test.pypi.org/manage/account/publishing/ and repeat

>    the second step, but now enter ``testpypi`` as the name of the

Reviewer: Let's avoid the repetition of "now":
```suggestion
   the second step, but this time, enter ``testpypi`` as the name of the
```

>    GitHub Actions environment.

Reviewer: ...and fix the terminology:
```suggestion
   GitHub Environment.
```

4. Your "pending" publishers are now ready for their first use and will
   create your projects automatically once you use them
   for the first time.

   .. note::

      If you don't have a TestPyPI account, you'll need to
      create it. It's not the same as a regular PyPI account.

   .. hint::

>       For security reasons, you should require manual approval

Reviewer: Many people probably wouldn't know what this is and where to set it up. We should either add this to the steps above or at least, link something like https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#deployment-protection-rules.

      create it. It's not the same as a regular PyPI account.

> Using trusted publishing

Reviewer: The idea of the guide is to be a simple and straightforward starting point. It should demonstrate a flow without any deviations, it should be linear.
Also, OIDC is more important as the default for anybody using GitHub. We can mention that it's possible to use login+password when publishing to third party indexes but it mustn't be the central part of this document.

Author: I agree and have adapted the documentation and release workflow file.

Reviewer: Marking this thread as resolved.

And now we can build the dists from source and store them.
In this example, we'll use the ``build`` package.
So add this to the steps list:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: version: "3.x"
   :end-before: publish-to-pypi

Defining a workflow job environment
===================================

Now, let's add initial setup for our job that will publish to PyPI.
It's a process that will execute commands that we'll define later.
In this guide, we'll use the latest stable Ubuntu LTS version
provided by GitHub Actions. This also defines the package index 
to publish to, PyPI, and grants a permission to the action that 
is mandatory for trusted publishing.

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: name: python-package-distributions
   :end-before: steps:

Publishing the distribution to PyPI
===================================

Finally, add the following steps at the end:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :lines: 41-48

This step uses the `pypa/gh-action-pypi-publish`_ GitHub

> Action: After the stored distribution package has been 

Reviewer: ```suggestion
Action: after the stored distribution package has been 
```

downloaded by the `download-artifact`_ action, it uploads 

> the contents of the ``dist/`` folder into PyPI unconditionally.

Reviewer: ```suggestion
the contents of the ``dist/`` folder into PyPI unconditionally.
This job also signs the artifacts with Sigstore right after publishing them to PyPI.
```

   :language: yaml
   :end-before: jobs:

> This will also ensure that the release workflow is only triggered

Reviewer 3: <!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏the updated comment states that this will only work on tagged, but the current yml example shows `on: push` - is that correct?

A later example shows `if: startsWith(github.ref, 'refs/tags/')` - that's only relevant to the testpypi workflow, not the pypi one.

Would it be better expressed as something like:

```yaml
on:
  push:
    tags:
      - '*'
```

Reviewer: No, this is an incorrect assumption — the workflow can run on normal pushes and the build step would be a smoke test while the publishing would be skipped. Honestly, I'd rather refactor this to trigger on `workflow_dispatch:` but that would be too complex — maybe it's better to have a separate guide for a more complicated / comprehensive scenario.
I usually expect this to be a part of the same main CI/CD workflow that also includes tests. This way, you can also test those built artifacts before publishing.

Reviewer 3: Okay - I'll rely on your experience here. I think I've seen the build/publish steps as part of a distinct workflow, which would trigger on tags, so if the intent to keep this all in one workflow, then the refs/tags approach is fine!

Reviewer: > Okay - I'll rely on your experience here. I think I've seen the build/publish steps as part of a distinct workflow, which would trigger on tags, so if the intent to keep this all in one workflow, then the refs/tags approach is fine!

Yeah, I've seen different workflows too. I just didn't want to change too many things in this PR. But later, I'd definitely rethink this. Though, I was considering a separate more in-depth document rather then extending this with multiple branches.

Reviewer: Marking this as resolved.

if the current commit is tagged. It is recommended you use the
latest release tag.

Checking out the project and building distributions
===================================================

We will have to define two jobs to publish to PyPI 
and TestPyPI respectively, and an additional job to 
build the distribution packages.

First, we'll define the job for building the dist packages of 
your project and storing them for later use:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml

>    :start-after: jobs:

Reviewer 3: <!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏this change makes `jobs:` never appear in the rendered document, so there's no place for the user to know that there needs to be a `jobs:` key.

Author: Thanks, maybe the `start-before` directive is better here.

   :end-before: Install pypa/build

This will download your repository into the CI runner and then
install and activate the newest available Python 3 release.

And now we can build the dists from source and store them.
In this example, we'll use the ``build`` package.
So add this to the steps list:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: version: "3.x"
   :end-before: publish-to-pypi

Defining a workflow job environment
===================================

Now, let's add initial setup for our job that will publish to PyPI.
It's a process that will execute commands that we'll define later.
In this guide, we'll use the latest stable Ubuntu LTS version
provided by GitHub Actions. This also defines a GitHub Environment
for the job to run in its context and a URL to be displayed in GitHub's
UI nicely. Additionally, it allows aqcuiring an OpenID Connect token
which is mandatory that the ``pypi-publish`` actions needs to
implement secretless trusted publishing to PyPI.

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: path: dist/

>    :end-before: steps:

Reviewer 3: This step shows the user the example `url: https://pypi.org/p/<package-name>` but makes no mention of which name to place in `package-name`

   :start-at: jobs:
   :end-before: Install pypa/build

This will download your repository into the CI runner and then
install and activate the newest available Python 3 release.

And now we can build the dists from source and store them.
In this example, we'll use the ``build`` package.
So add this to the steps list:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-at: Install pypa/build
   :end-before: publish-to-pypi

Defining a workflow job environment
===================================

Now, let's add initial setup for our job that will publish to PyPI.
It's a process that will execute commands that we'll define later.
In this guide, we'll use the latest stable Ubuntu LTS version
provided by GitHub Actions. This also defines a GitHub Environment
for the job to run in its context and a URL to be displayed in GitHub's
UI nicely. Additionally, it allows aqcuiring an OpenID Connect token
that the ``pypi-publish`` actions needs to implement secretless
trusted publishing to PyPI.

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: path: dist/
   :end-before: steps:

> This will also ensure that the PyPI publishing workflow is only triggered

Reviewer: Let's also mention somewhere in this paragraph that it's important to set up the manual approval requirement for the `pypi` gh env.

if the current commit is tagged. It is recommended you use the

> latest release tag.

Reviewer: I'm not sure what this last sentence means. Can we rephrase it to make it more illustrative?

Publishing the distribution to PyPI
===================================

Finally, add the following steps at the end:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: id-token: write
   :end-before:  github-release:

This step uses the `pypa/gh-action-pypi-publish`_ GitHub
Action: after the stored distribution package has been
downloaded by the `download-artifact`_ action, it uploads
the contents of the ``dist/`` folder into PyPI unconditionally.

Signing the distribution packages
=================================

This additional job signs the distribution packages with `Sigstore`_,
using the `sigstore/gh-action-sigstore-python GitHub Action`_,

> and then uploads them to GitHub Release.

Reviewer: ```suggestion
and then uploads them to a GitHub Release.
```

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-at: github-release:
   :end-before:  publish-to-testpypi

.. note::

   This is a replacement for GPG signatures, for which support has been

>    `removed <https://blog.pypi.org/posts/2023-05-23-removing-pgp/>`_ by PyPI.

Reviewer: how about
```suggestion
   `removed from PyPI <https://blog.pypi.org/posts/2023-05-23-removing-pgp/>`_.
```

>    However, this job is not mandatory for defining the workflow.

Reviewer: ```suggestion
   However, this job is not mandatory for uploading to PyPI and can be omitted.
```

Separate workflow for publishing to TestPyPI
============================================

Now, repeat these steps and create another job for
publishing to the TestPyPI package index under the ``jobs``
section:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-at: publish-to-testpypi

> 

Reviewer: ```suggestion

.. tip::

   Requiring manual approvals in the `testpypi` GitHub Environment is typically unnecessary as it's designed to run on each commit to the main branch and is often used to indicate a healthy release publishing pipeline.

```

And now we can build the dists from source and store them.
In this example, we'll use the ``build`` package.
So add this to the steps list:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-at: Install pypa/build
   :end-before: publish-to-pypi

Defining a workflow job environment
===================================

Now, let's add initial setup for our job that will publish to PyPI.
It's a process that will execute commands that we'll define later.
In this guide, we'll use the latest stable Ubuntu LTS version
provided by GitHub Actions. This also defines a GitHub Environment
for the job to run in its context and a URL to be displayed in GitHub's
UI nicely. Additionally, it allows aqcuiring an OpenID Connect token
that the ``pypi-publish`` actions needs to implement secretless
trusted publishing to PyPI.

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: path: dist/
   :end-before: steps:

This will also ensure that the PyPI publishing workflow is only triggered
if the current commit is tagged. It is recommended that you commit using the

> latest release tag.

Author: I haven't added this; see https://github.com/pypa/packaging.python.org/pull/1092#discussion_r929321386 for context. I see you have been struggling with that wording back then, too ;-). There is an unaccepted suggestion in the same discussion that could be added.

But I agree I understood it falsely at first, which is why the sentence is even more misleading now. 

I'm quite sure this means keeping the release tag of _gh-action-pypi-publish_ up-to-date (compare with the current version of the guide) . But since we have quite a lot of actions integrated now, I think we should either take this out or move it to a more obvious position.

> latest release tag.

Reviewer: Oh, that didn't occur to me at all. Agreed, moving and making it clear that it's about the actions that are integrated is a good idea.

In this guide, we'll use the latest stable Ubuntu LTS version
provided by GitHub Actions. This also defines a GitHub Environment
for the job to run in its context and a URL to be displayed in GitHub's
UI nicely. Additionally, it allows aqcuiring an OpenID Connect token
that the ``pypi-publish`` actions needs to implement secretless
trusted publishing to PyPI.

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: path: dist/
   :end-before: steps:

This will also ensure that the PyPI publishing workflow is only triggered
if the current commit is tagged. It is recommended you use the
latest release tag.

Publishing the distribution to PyPI
===================================

Finally, add the following steps at the end:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-after: id-token: write
   :end-before:  github-release:

This step uses the `pypa/gh-action-pypi-publish`_ GitHub
Action: after the stored distribution package has been
downloaded by the `download-artifact`_ action, it uploads
the contents of the ``dist/`` folder into PyPI unconditionally.

> Signing the distribution packages

Reviewer 4: Do we want to have Sigstore signing in the guide for publishing with Trusted Publishers? These are different concepts and the current reading makes it seem like you need both to accomplish the task?

Reviewer: @sethmlarson this is not a tutorial about trusted publishing but about publishing to PyPI from GHA which happens to use OIDC instead of long-lived credentials now. Though, I get your point and it's probably hard to pick where to draw the line between what's in and out of the scope. Perhaps, a `.. note::` admonition could explain that this job is not mandatory.
But I do feel like it's important to show people a friction-free way of signing their artifacts, especially given the now-official deprecation of GPGs in Warehouse.

=================================

This additional job signs the distribution packages with the
`sigstore/gh-action-sigstore-python GitHub Action`_ and then uploads
them to GitHub Release.

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-at: github-release:
   :end-before:  publish-to-testpypi

Separate workflow for publishing to TestPyPI
============================================

Now, repeat these steps and create another job for
publishing to the TestPyPI package index under the ``jobs``
section:

.. literalinclude:: github-actions-ci-cd-sample/publish-to-test-pypi.yml
   :language: yaml
   :start-at: publish-to-testpypi

> The whole CD workflow

Reviewer: This section looks weird with just one collapsed line. Can we add some paragraph at the beginning? Or think of some other structure?

=====================

> .. collapse:: Load file

Reviewer: ```suggestion
.. collapse:: Click here to display the entire GitHub Actions CI/CD workflow definition
```

   https://github.com/marketplace/actions/pypi-publish
.. _Secrets:
   https://docs.github.com/en/actions/reference/encrypted-secrets

> .. _trusted publishing: https://docs.pypi.org/trusted-publishers/

Author: The PyPI docs haven't configured that. They'd need the [mkdocstrings](https://mkdocstrings.github.io/usage/#cross-references-to-other-projects-inventories) plugin.

I consider adding it there.

> .. _trusted publishing: https://docs.pypi.org/trusted-publishers/

Reviewer: I'm not sure. This doc describes referencing other inventories, not producing one..

> .. _trusted publishing: https://docs.pypi.org/trusted-publishers/

Author: See the last example.

> .. _trusted publishing: https://docs.pypi.org/trusted-publishers/

Reviewer: Oh, I see now. Okay.

> .. _trusted publishing: https://docs.pypi.org/trusted-publishers/

Reviewer: Marking this thread as resolved.
